### ISSUE TITLE
-ldl need by silo H5open conftest, otherwise H5PL.c:(.text+0xb5): undefined reference to `dlclose'

### ISSUE THREAD

**Issue by pozulp** [spack/spack#5453]
@becker33 

Executive summary
===============
HDF5 1.8.11 added dynamical loading of HDF5 filters and therefore depends on libdl. When I try to build silo, which depends on hdf5, silo's configure fails on a conftest for H5open. If you google "hdf5 dlopen" you find [this](http://hdf-forum.184993.n3.nabble.com/Errors-compiling-against-Static-build-HDF5-1-8-11-Need-for-ldl-added-to-linker-arguments-td4026300.html) which says add "-ldl" to your silo ldflags. Is this what I want, and if so how should I make spack add the ldflag?

Note
-------
I tried building silo with hdf5 1.8.10 (predates libdl dependency) but this fails with a compile error. I will report this in a separate issue (see #5454). The workaround may be as simple as adding a patch for hdf5 1.8.10 that patches the src causing the compile error. I also tried building silo with hdf5 1.10.1, the "Preferred version" according to spack info hdf5.

My runline and output
==================
When I run
`$ ./bin/spack install --keep-stage silo@4.10.2%gcc@4.8.5~fortran+pic~shared~silex arch=linux-rhel7-x86_64 ^hdf5@1.8.12%gcc@4.8.5~cxx~debug~fortran~mpi+pic~shared~szip~threadsafe arch=linux-rhel7-x86_64`
I get the following

==> zlib is already installed in /home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/zlib-1.2.11-hpdicdxfq5jn2ycttiuiu42hemg444nf
==> Installing hdf5
==> Using cached archive: /home/pozulp1/packages/spack/var/spack/cache/hdf5/hdf5-1.8.12.tar.gz
==> Already staged hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6 in /home/pozulp1/packages/spack/var/spack/stage/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6
==> No patches needed for hdf5
==> Building hdf5 [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed hdf5
  Fetch: 0.02s.  Build: 1m 5.59s.  Total: 1m 5.61s.
[+] /home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6
==> Installing silo
==> Fetching https://wci.llnl.gov/content/assets/docs/simulation/computer-codes/silo/silo-4.10.2/silo-4.10.2.tar.gz
######################################################################## 100.0%
==> Staging archive: /home/pozulp1/packages/spack/var/spack/stage/silo-4.10.2-ovc4qkobypm6asdydhtt3eqr4pu2ivcp/silo-4.10.2.tar.gz
==> Created stage in /home/pozulp1/packages/spack/var/spack/stage/silo-4.10.2-ovc4qkobypm6asdydhtt3eqr4pu2ivcp
==> Applied patch remove-mpiposix.patch
==> Building silo [Package]
==> Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    './configure' '--prefix=/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/silo-4.10.2-ovc4qkobypm6asdydhtt3eqr4pu2ivcp' '--with-hdf5=/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6/include,/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6/lib' '--with-zlib=/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/zlib-1.2.11-hpdicdxfq5jn2ycttiuiu42hemg444nf/include,/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/zlib-1.2.11-hpdicdxfq5jn2ycttiuiu42hemg444nf/lib' '--enable-install-lite-headers' '--disable-fortran' '--disable-silex' '--disable-shared' 'CFLAGS=-fPIC' 'CXXFLAGS=-fPIC' 'FCFLAGS=-fPIC'

1 error found in build log:
     [ ... ]
     182   checking for H5open... no
     183   configure: checking if lack of zlib caused tests for hdf5 to fail...
     184   checking hdf5.h usability... yes
     185   checking hdf5.h presence... yes
     186   checking for hdf5.h... yes
     187   checking for H5open in -lhdf5... no
  >> 188   configure: error:
     189   You have made an explicit request for hdf5 with the --with-hdf option
     190   and ./configure is unable to find hdf5 and its supporting libs, if any.
     191   Please see INSTALL_FAQ for possible causes.

See build log for details:
  /home/pozulp1/packages/spack/var/spack/stage/silo-4.10.2-ovc4qkobypm6asdydhtt3eqr4pu2ivcp/silo-4.10.2/spack-build.out

The contents of silo's config.log
=======================
[pozulp1@bubba spack]$ grep -A 10 "checking for H5open in" /home/pozulp1/packages/spack/var/spack/stage/silo-4.10.2-ovc4qkobypm6asdydhtt3eqr4pu2ivcp/silo-4.10.2/config.log 
configure:29453: checking for H5open in -lhdf5
configure:29496: /home/pozulp1/packages/spack/lib/spack/env/gcc/gcc -o conftest -fPIC -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Wdeclaration-after-statement -I/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6/include  -I/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/zlib-1.2.11-hpdicdxfq5jn2ycttiuiu42hemg444nf/include -L/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6/lib  -L/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/zlib-1.2.11-hpdicdxfq5jn2ycttiuiu42hemg444nf/lib conftest.c -lhdf5  -lz  -lz -lm >&5
/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6/lib/libhdf5.a(H5PL.o): In function `H5PL_term_interface':
H5PL.c:(.text+0xb5): undefined reference to `dlclose'
/home/pozulp1/packages/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/hdf5-1.8.12-lt6wej75hjdpliylqib3sb3ouidh7jr6/lib/libhdf5.a(H5PL.o): In function `H5PL_load':
H5PL.c:(.text+0x35e): undefined reference to `dlsym'
H5PL.c:(.text+0x4ac): undefined reference to `dlopen'
H5PL.c:(.text+0x4c7): undefined reference to `dlsym'
H5PL.c:(.text+0x614): undefined reference to `dlclose'
H5PL.c:(.text+0x699): undefined reference to `dlerror'
H5PL.c:(.text+0x902): undefined reference to `dlclose'
[pozulp1@bubba spack]$

**Comment by omsai** [comment]
A minimum reproducible example of not finding H5open in -lhdf5:

```console
$ spack install silo~shared ^hdf5~shared
...
==> Installing silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5
==> No binary for silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5 found: installing from source
==> Using cached archive: /home/omsai/src/spack/var/spack/cache/_source-cache/archive/3a/3af87e5f0608a69849c00eb7c73b11f8422fa36903dd14610584506e7f68e638.tgz
==> Applied patch /home/omsai/src/spack/var/spack/repos/builtin/packages/silo/remove-mpiposix.patch
==> Applied patch /home/omsai/src/spack/var/spack/repos/builtin/packages/silo/H5FD_class_t-terminate.patch
==> Applied patch /home/omsai/src/spack/var/spack/repos/builtin/packages/silo/H5EPR_SEMI_COLON.patch
==> Applied patch /home/omsai/src/spack/var/spack/repos/builtin/packages/silo/configure-AX_CHECK_COMPILE_FLAG.patch
==> silo: Executing phase: 'autoreconf'
==> silo: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/omsai/spack-stage/spack-stage-silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5/spack-src/configure' '--prefix=/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5' '--enable-install-lite-headers' '--enable-fortran' '--disable-silex' '--disable-shared' '--enable-hzip' '--enable-fpzip' '--with-zlib=/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/include,/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/lib' '--with-hdf5=/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/include,/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib' 'CC=/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/openmpi-4.1.4-xing6mvnayn234mgstxocmkjulitrg4l/bin/mpicc' 'CXX=/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/openmpi-4.1.4-xing6mvnayn234mgstxocmkjulitrg4l/bin/mpic++' 'FC=/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/openmpi-4.1.4-xing6mvnayn234mgstxocmkjulitrg4l/bin/mpif90'

1 error found in build log:
     194    checking for H5open... no
     195    configure: checking if lack of zlib caused tests for hdf5 to fail...
     196    checking hdf5.h usability... yes
     197    checking hdf5.h presence... yes
     198    checking for hdf5.h... yes
     199    checking for H5open in -lhdf5... no
  >> 200    configure: error:
     201    You have made an explicit request for hdf5 with the --with-hdf option
     202    and /tmp/omsai/spack-stage/spack-stage-silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5/spack-src/configure is unable to find hdf5 and its supporting libs, if any.
     203    Please see INSTALL_FAQ for possible causes.

See build log for details:
  /tmp/omsai/spack-stage/spack-stage-silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5/spack-build-out.txt
```

As you say, the underlying error in `config.log` is `undefined reference to 'dlsym'`

```console
$ awk '/checking for H5open in -lhdf5/ {flag=1} flag==1 && $0 ~ /^$/ {flag=0} flag==1 {print}'  /tmp/omsai/spack-stage/spack-stage-silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5/spack-src/config.log
configure:29437: checking for H5open in -lhdf5
configure:29480: /home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/openmpi-4.1.4-xing6mvnayn234mgstxocmkjulitrg4l/bin/mpicc -o conftest -fPIC -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Wdeclaration-after-statement -I/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/include  -I/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/include -L/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib  -L/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/lib conftest.c -lhdf5  -lz  -lz -lm >&5
/usr/bin/ld: /home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib/libhdf5.a(H5PLint.c.o): in function `H5PL__close':
/tmp/omsai/spack-stage/spack-stage-hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/spack-src/src/H5PLint.c:431: undefined reference to `dlclose'
/usr/bin/ld: /home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib/libhdf5.a(H5PLint.c.o): in function `H5PL__open':
/tmp/omsai/spack-stage/spack-stage-hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/spack-src/src/H5PLint.c:321: undefined reference to `dlopen'
/usr/bin/ld: /tmp/omsai/spack-stage/spack-stage-hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/spack-src/src/H5PLint.c:329: undefined reference to `dlsym'
/usr/bin/ld: /tmp/omsai/spack-stage/spack-stage-hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/spack-src/src/H5PLint.c:335: undefined reference to `dlsym'
/usr/bin/ld: /tmp/omsai/spack-stage/spack-stage-hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/spack-src/src/H5PLint.c:322: undefined reference to `dlerror'
/usr/bin/ld: /home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib/libhdf5.a(H5PLplugin_cache.c.o): in function `H5PL__find_plugin_in_cache':
/tmp/omsai/spack-stage/spack-stage-hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/spack-src/src/H5PLplugin_cache.c:275: undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
configure:29487: $? = 1
configure: failed program was:
| /* confdefs.h.  */
| #define PACKAGE_NAME "silo"
| #define PACKAGE_TARNAME "silo"
| #define PACKAGE_VERSION "4.10.2"
| #define PACKAGE_STRING "silo 4.10.2"
| #define PACKAGE_BUGREPORT "visit-users@elist.ornl.gov"
| #define PACKAGE "silo"
| #define VERSION "4.10.2"
| #define FC_FUNC(name,NAME) name ## _
| #define FC_FUNC_(name,NAME) name ## _
| #define STDC_HEADERS 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_SYS_STAT_H 1
| #define HAVE_STDLIB_H 1
| #define HAVE_STRING_H 1
| #define HAVE_MEMORY_H 1
| #define HAVE_STRINGS_H 1
| #define HAVE_INTTYPES_H 1
| #define HAVE_STDINT_H 1
| #define HAVE_UNISTD_H 1
| #define HAVE_DLFCN_H 1
| #define LT_OBJDIR ".libs/"
| #define NORMAL_SCLITE_HEADERS 0
| #define HAVE_NETCDF_DRIVER 1
| #define HAVE_PDB_DRIVER 1
| #define HAVE_TAURUS_DRIVER 1
| #define HAVE_ZLIB_H 1
| #define HAVE_LIBZ 1
| #define HAVE_HZIP 1
| #define HAVE_ZLIB_H 1
| #define HAVE_LIBZ 1
| #define HAVE_FPZIP 1
| #define SIZEOF_OFF64_T 8
| #define HAVE_SYS_TIME_H 1
| #define HAVE_STDARG_H 1
| #define HAVE_SYS_STAT_H 1
| #define HAVE_FCNTL_H 1
| #define HAVE_SYS_FCNTL_H 1
| #define HAVE_FNMATCH_H 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_UNISTD_H 1
| #define STDC_HEADERS 1
| #define HAVE_FCLOSE_POINTER 1
| #define HAVE_FFLUSH_POINTER 1
| #define HAVE_FOPEN_POINTER 1
| #define HAVE_FPRINTF_POINTER 1
| #define HAVE_FREAD_POINTER 1
| #define HAVE_FSEEK_POINTER 1
| #define HAVE_SETVBUF_POINTER 1
| #define HAVE_FTELL_POINTER 1
| #define HAVE_FWRITE_POINTER 1
| #define HAVE_MEMMOVE 1
| #define HAVE_FNMATCH 1
| #define HAVE_ISNAN 1
| #define HAVE_STRERROR 1
| #define HAVE_HDF5_H 1
| #define HAVE_HDF5_H 1
| /* end confdefs.h.  */
| 
| /* Override any GCC internal prototype to avoid an error.
|    Use char because int might match the return type of a GCC
|    builtin and then its argument prototype would still apply.  */
| #ifdef __cplusplus
| extern "C"
| #endif
| char H5open ();
| #ifdef FC_DUMMY_MAIN
| #ifndef FC_DUMMY_MAIN_EQ_F77
| #  ifdef __cplusplus
|      extern "C"
| #  endif
|    int FC_DUMMY_MAIN() { return 1; }
| #endif
| #endif
| int
| main ()
| {
| return H5open ();
|   ;
|   return 0;
| }
configure:29508: result: no
configure:29537: error:
You have made an explicit request for hdf5 with the --with-hdf option
and /tmp/omsai/spack-stage/spack-stage-silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5/spack-src/configure is unable to find hdf5 and its supporting libs, if any.
Please see INSTALL_FAQ for possible causes.
```

Note that the second line in the output above does not use `-ldl`:

```console
$ awk '/29480/' /tmp/omsai/spack-stage/spack-stage-silo-4.10.2-nylp7b5664srujpa6qram3bqbi2tsbf5/spack-src/config.log | sed -E 's#[[:space:]]+# #g' | tr ' ' '\n'
configure:29480:
/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/openmpi-4.1.4-xing6mvnayn234mgstxocmkjulitrg4l/bin/mpicc
-o
conftest
-fPIC
-D_LARGEFILE_SOURCE
-D_LARGEFILE64_SOURCE
-D_FILE_OFFSET_BITS=64
-Wdeclaration-after-statement
-I/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/include
-I/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/include
-L/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib
-L/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/lib
conftest.c
-lhdf5
-lz
-lz
-lm
>&5
```

This hack currently in `silo` adds `-ldl` but clearly it's not propagating to the configure test program and needs to be fixed:

```python
class Silo(AutotoolsPackage):
...
    def flag_handler(self, name, flags):
        spec = self.spec
        if name == "ldflags":
            if "+hdf5" in spec:
                if spec["hdf5"].satisfies("~shared"):
                    flags.append("-ldl")
            flags.append(spec["readline"].libs.search_flags)
```

**Comment by omsai** [comment]
It's really bizarre that `-dl` isn't being added in the configure test compilation command but the rest of the static libs are, because hdf5's pkg-config with `--static` provides `-dl`:

```console
$ spack unload --all
$ spack load hdf5~shared
$ pkg-config --libs hdf5
-L/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib
$ pkg-config --static --libs hdf5 | tr ' ' '\n'
-L/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/hdf5-1.12.2-p5hdbkf4yoyogro4oe5mfjykdqpiyig7/lib
-lm
-ldl
-l/home/omsai/src/spack/opt/spack/linux-debian11-nehalem/gcc-10.2.1/zlib-1.2.12-fl5xq6mkm26z7zhlzqgr4tayy4g57lab/lib/libz.so
-lhdf5
